server/server.go

```diff
@@ -59,28 +64,70 @@ func (s *Server) Run(svc transport.Service) error {
 }
 
 func (s *Server) run() error {
-	errChan := make(chan error, 5)
-
+	var g errgroup.Group
 	if s.listeners.mainListener != nil {
-		go func() {
+		g.Go(func() error {
 			err := s.listeners.mainListener.Serve()
-			errChan <- err
-		}()
+			return err
+		})
+	}
+	sigint := make(chan os.Signal, 1)
+	signal.Notify(sigint, os.Interrupt)
+	if s.opts.GracefullFunc != nil {
+		fn := s.opts.GracefullFunc(sigint)
+		g.Go(fn)
+	} else {
+		g.Go(func() error {
+			<-sigint
+			s.Stop()
+			return nil
+		})
 	}
-	go func() {
+	g.Go(func() error {
+		if s.opts.HTTPServer != nil {
+			s.opts.HTTPServer.Handler = s.srv.http
+			err := s.opts.HTTPServer.Serve(s.listeners.HTTP)
+			return err
+		}
 		err := http.Serve(s.listeners.HTTP, s.srv.http)
-		errChan <- err
-	}()
-	go func() {
+		return err
+	})
+	g.Go(func() error {
 		err := s.srv.grpc.Serve(s.listeners.GRPC)
-		errChan <- err
-	}()
+		return err
+	})
 
-	return <-errChan
+	return g.Wait()
 }
 
 // Stop stops the server gracefully.
 func (s *Server) Stop() {
-	// TODO grace HTTP
-	s.srv.grpc.GracefulStop()
+	var g errgroup.Group
+	if s.opts.HTTPServer != nil {
+		g.Go(func() error {
+			return s.opts.HTTPServer.Shutdown(context.Background())
```

Comment:
It's not proper `GracefulStop` since we don't have any timeout provided to `Shutdown()`, potentially hanging the app indefinitely. `Stop()` should work as it works now